### PR TITLE
Improve coverage

### PR DESCRIPTION
**Pull request by tkhurana96**

### CODE REVIEW

File: examples/socket_unix_udp_client.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+int main()
+{
+	try {
+		Socket unixClient(Domain::UNIX, Type::UDP);
+		std::string clientPath("/tmp/unixClient");
+		std::string serverPath("/tmp/unixServer");
+
+		unixClient.bind([&](AddrUnix &s) {
+			return methods::construct(s, &clientPath.front());
```
Comment (agauniyal) [inline examples/socket_unix_udp_client.cpp]: why not `.c_str()`?

File: examples/socket_unix_udp_client.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+		});
+
+		unixClient.connect(&serverPath.front());
```
Comment (agauniyal) [inline examples/socket_unix_udp_client.cpp]: `.c_str()`?

File: examples/socket_unix_udp_server.cpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+using namespace net;
+
+int main()
+{
+	try {
+		Socket unixServer(Domain::UNIX, Type::UDP);
+		std::string serverPath("/tmp/unixServer");
+
+		unixServer.bind([&](AddrUnix &s) {
+			return methods::construct(s, &serverPath.front());
```
Comment (agauniyal) [inline examples/socket_unix_udp_server.cpp]: `.c_str()`?

File: test/socket_connect_test.cpp [hunk 4]
```diff
@@ -45,11 +54,50 @@ TEST(Socket, Connectv4)
...
 }
+
+TEST(Socket, ConnectUnixTCP)
+{
+	std::string unixPathServer("/tmp/unixSocketFileServer7");
+	Socket serverUnixTCP(Domain::UNIX, Type::TCP);
+	std::thread serverThread1(runUnixServer, std::ref(serverUnixTCP),
+	                          &unixPathServer.front());
+	serverThread1.detach();
+	std::this_thread::sleep_for(2s);
```
Comment (agauniyal) [inline test/socket_connect_test.cpp]: try lowering `2s` delay if possible.

File: test/socket_constructor_test.cpp [hunk 5]
```diff
@@ -13,6 +13,9 @@ TEST(Socket, ConstructorIPv4)
 
 	EXPECT_ANY_THROW(Socket s(Domain::IPv4, Type::TCP, 41));
 	EXPECT_ANY_THROW(Socket s(Domain::IPv4, Type::UDP, 6));
+
+	EXPECT_NO_THROW(Socket otherS(Socket(Domain::IPv4, Type::TCP)));
```
Comment (agauniyal) [inline test/socket_constructor_test.cpp:17]: What is this?

File: test/socket_send_test.cpp [hunk 6]
```diff
@@ -93,15 +85,15 @@ void startTCPServerIPv4()
 void startUDPServerIPv6()
 {
 	Socket myServerSocket(Domain::IPv6, Type::UDP);
-	myServerSocket.start("::1", 16000);
+	myServerSocket.start("::1", 16000);  // just binding
```
Comment (agauniyal) [inline test/socket_send_test.cpp]: yes this will just bind because `listen` only works for `Connections`.

File: test/socket_send_test.cpp [hunk 7]
```diff
@@ -93,15 +85,15 @@ void startTCPServerIPv4()
...
 	udpIPv6ServerProcessing(myServerSocket);
 }
 
 
 void startUDPServerIPv4()
 {
 	Socket myServerSocket(Domain::IPv4, Type::UDP);
-	myServerSocket.start("127.0.0.1", 15000);
+	myServerSocket.start("127.0.0.1", 15000);  // just binding
```
Comment (agauniyal) [inline test/socket_send_test.cpp]: yes this will just bind because `listen` only works for `Connections`.

File: test/socket_connect_test.cpp [hunk 8]
```diff
@@ -45,11 +54,50 @@ TEST(Socket, Connectv4)
...
 	ASSERT_NO_THROW(client6.connect("::1", 15020));
 	client6.close();
 }
+
+TEST(Socket, ConnectUnixTCP)
+{
+	std::string unixPathServer("/tmp/unixSocketFileServer7");
+	Socket serverUnixTCP(Domain::UNIX, Type::TCP);
+	std::thread serverThread1(runUnixServer, std::ref(serverUnixTCP),
+	                          &unixPathServer.front());
```
Comment (agauniyal) [inline test/socket_connect_test.cpp]: `.c_str()`?

File: test/socket_connect_test.cpp [hunk 9]
```diff
@@ -45,11 +54,50 @@ TEST(Socket, Connectv4)
...
+	serverThread1.detach();
+	std::this_thread::sleep_for(2s);
+
+	std::string unixPathClient("/tmp/unixSocketFileClient7");
+	Socket clientUnixTCP(Domain::UNIX, Type::TCP);
+	EXPECT_NO_THROW(clientUnixTCP.bind([&](AddrUnix &s) {
+		return methods::construct(s, &unixPathClient.front());
```
Comment (agauniyal) [inline test/socket_connect_test.cpp]: `.c_str()`?

File: test/socket_connect_test.cpp [hunk 10]
```diff
@@ -45,11 +54,50 @@ TEST(Socket, Connectv4)
...
+	}));
+	EXPECT_NO_THROW(clientUnixTCP.connect([&](AddrUnix &s) {
+		return methods::construct(s, &unixPathServer.front());
+	}););
+	clientUnixTCP.close();
```
Comment (agauniyal) [inline test/socket_connect_test.cpp]: shouldn't need to explicitly call `close()` here, ending `}` is on next line only.

File: test/socket_connect_test.cpp [hunk 11]
```diff
@@ -45,11 +54,50 @@ TEST(Socket, Connectv4)
...
+}
+
+TEST(Socket, ConnectUnixUDP)
+{
+	std::string unixPathServer("/tmp/unixSocketFileServer2");
+	Socket serverUnixUDP(Domain::UNIX, Type::UDP);
+	std::thread serverThread1(
+	  [&]() { runUnixServer(serverUnixUDP, &unixPathServer.front()); });
```
Comment (agauniyal) [inline test/socket_connect_test.cpp]: `.c_str()`?

File: test/socket_connect_test.cpp [hunk 12]
```diff
@@ -45,11 +54,50 @@ TEST(Socket, Connectv4)
...
+	serverThread1.detach();
+	std::this_thread::sleep_for(1s);
+
+	std::string unixPathClient("/tmp/unixSocketFileClient2");
+	Socket clientUnixUDP(Domain::UNIX, Type::UDP);
+	EXPECT_NO_THROW(clientUnixUDP.bind([&](AddrUnix &s) {
+		return methods::construct(s, &unixPathClient.front());
+	}));
+	EXPECT_NO_THROW(clientUnixUDP.connect([&](AddrUnix &s) {
+		return methods::construct(s, &unixPathServer.front());
```
Comment (agauniyal) [inline test/socket_connect_test.cpp]: `.c_str()`?

File: test/socket_connect_test.cpp [hunk 13]
```diff
@@ -45,11 +54,50 @@ TEST(Socket, Connectv4)
...
+	Socket serverUnixUDP(Domain::UNIX, Type::UDP);
+	std::thread serverThread1(
+	  [&]() { runUnixServer(serverUnixUDP, &unixPathServer.front()); });
+	serverThread1.detach();
+	std::this_thread::sleep_for(1s);
+
+	std::string unixPathClient("/tmp/unixSocketFileClient2");
+	Socket clientUnixUDP(Domain::UNIX, Type::UDP);
+	EXPECT_NO_THROW(clientUnixUDP.bind([&](AddrUnix &s) {
+		return methods::construct(s, &unixPathClient.front());
```
Comment (agauniyal) [inline test/socket_connect_test.cpp]: `.c_str()`?

File: test/socket_read_write_test.cpp [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+	  [&](AddrUnix &s) { return methods::construct(s, &path.front()); }));
+
+	const auto res = unixServer.read(readTest::msgLen);
+	EXPECT_EQ(msg, res);
+}
+
+void udpIPv4ServerProcessing(Socket &serverSocket)
+{
+	const auto res = serverSocket.read(readTest::msgLen);
+	EXPECT_EQ(res, readTest::msg);
```
Comment (agauniyal) [inline test/socket_read_write_test.cpp:47]: Why using `EXPECT_EQ`?

File: test/socket_read_write_test.cpp [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+void startUNIXServerUDP(const std::string path)
+{
+	Socket unixServer(Domain::UNIX, Type::UDP);
+	EXPECT_NO_THROW(unixServer.bind(
+	  [&](AddrUnix &s) { return methods::construct(s, &path.front()); }));
+
+	const auto res = unixServer.read(readTest::msgLen);
+	EXPECT_EQ(msg, res);
```
Comment (agauniyal) [inline test/socket_read_write_test.cpp]: Why using `EXPECT_EQ`?

File: test/socket_read_write_test.cpp [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+void udpIPv4ServerProcessing(Socket &serverSocket)
+{
+	const auto res = serverSocket.read(readTest::msgLen);
+	EXPECT_EQ(res, readTest::msg);
+}
+
+void udpIPv6ServerProcessing(Socket &serverSocket)
+{
+	const auto res = serverSocket.read(readTest::msgLen);
+	EXPECT_EQ(res, readTest::msg);
```
Comment (agauniyal) [inline test/socket_read_write_test.cpp:53]: Why using `EXPECT_EQ`?

File: test/socket_read_write_test.cpp [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+	std::thread tcpServerThreadUnix(readTest::startUNIXServerTCP,
+	                                std::ref(readTest::unixServerPath1));
+	std::thread udpServerThreadUnix(readTest::startUNIXServerUDP,
+	                                std::ref(readTest::unixServerPath2));
+	std::this_thread::sleep_for(1s);
+
+	Socket unixClient1(Domain::UNIX, Type::TCP);
+	EXPECT_EQ(unixClient1.getType(), Type::TCP);
+	EXPECT_NO_THROW(unixClient1.bind([&](AddrUnix &s) {
+		return methods::construct(s, &readTest::unixClientPath1.front());
```
Comment (agauniyal) [inline test/socket_read_write_test.cpp]: `.c.str()`?

File: test/socket_read_write_test.cpp [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+	}));
+	EXPECT_NO_THROW(unixClient1.connect(&readTest::unixServerPath1.front()));
```
Comment (agauniyal) [inline test/socket_read_write_test.cpp]: `.c.str()`?

File: test/socket_read_write_test.cpp [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+
+	EXPECT_NO_THROW(unixClient1.write(readTest::msg));
+	EXPECT_EQ(unixClient1.read(4), "readTest::msg");
+
+	unixClient1.close();
+
+	Socket unixClient2(Domain::UNIX, Type::UDP);
+	EXPECT_EQ(unixClient2.getType(), Type::UDP);
+	EXPECT_NO_THROW(unixClient2.bind([&](AddrUnix &s) {
+		return methods::construct(s, &readTest::unixClientPath2.front());
```
Comment (agauniyal) [inline test/socket_read_write_test.cpp]: `.c.str()`?

File: test/socket_read_write_test.cpp [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+	}));
+	EXPECT_NO_THROW(unixClient2.connect(&readTest::unixServerPath2.front()));
```
Comment (agauniyal) [inline test/socket_read_write_test.cpp]: `.c.str()`?

**Review by agauniyal — CHANGES_REQUESTED** [review]

File: test/socket_constructor_test.cpp (hunk 5 above)
Comment (tkhurana96) [reply test/socket_constructor_test.cpp:17]: trying the constructor which takes an rvalue

Comment (agauniyal) [reply test/socket_constructor_test.cpp:17]: @tkhurana96 try with a `vector.emplace()` kind of method then, which is more practical example.

**Review by agauniyal — APPROVED** [review]

**Pull request merged**